### ISSUE TITLE
OS X macOS endpoint and certificate error

### ISSUE THREAD

**Issue by baro77** [rclone/rclone#1150]
I'm using rclone 1.35 on macOS Sierra 10.12.3 with Amazon Cloud Drive remote. Sometimes  when I type a rclone command (can be the 'rclone mount', but also "legacy" ones as 'rclone check' and stuff like that) I get this error:

failed to get endpoints: Get https://drive.amazonaws.com/drive/v1/account/endpoint: x509: certificate signed by unknown authority

Usually retyping the command the problem doesn't happen.

Here it is an example with -vv option (unluckily not a lot of information, it seems):

Andreas-MacBook-Pro:~ baro$ rclone -vv copy -c --dry-run /Users/baro/mnt/COW/pippo2.sparsebundle/ ACD:/rclone/DATA/pippo2.sparsebundle/
2017/02/18 22:08:46 rclone: Version "v1.35" starting with parameters ["rclone" "-vv" "copy" "-c" "--dry-run" "/Users/baro/mnt/COW/pippo2.sparsebundle/" "ACD:/rclone/DATA/pippo2.sparsebundle/"]
2017/02/18 22:08:47 Failed to create file system for "ACD:/rclone/DATA/pippo2.sparsebundle/": failed to get endpoints: Get https://drive.amazonaws.com/drive/v1/account/endpoint: x509: certificate signed by unknown authority

**Comment by hashbackup** [comment]
I get a similar error trying to use rclone to access Google Drive from a Linode VM:
```
# rclone ls gdrive:
2017/02/16 16:37:31 Failed to create file system for "gdrive:":
couldn't read info about Drive: Get
https://www.googleapis.com/drive/v2/about?alt=json: Post
https://accounts.google.com/o/oauth2/token: x509: certificate signed
by unknown authority
```

I copied the .rclone.conf file from an OSX system where  it worked fine.  I tried deleting .rclone.conf and re-creating it on Linode, getting a new oauth authorization (that all worked), but still get the same error.  It doesn't go away when repeating the command.

I thought maybe Google doesn't like Linode's host key signatures?  Didn't spend any time trying to figure it out, just a guess.

**Comment by ncw** [comment]
@baro77 Try the steps in [this link](https://support.globalsign.com/customer/portal/articles/1353318-view-and-or-delete-crl-ocsp-cache) that has helped other mac users in the past.

@hashbackup you might need to update the ssl root certificates (that would be ca-certificates package on ubuntu).

I wonder if google have changed their certificates somehow recently...

**Comment by baro77** [comment]
@ncw thanks. I tried the suggestions in the link, but unluckily it wasn't really helpful... it happened again a couple of times after I had done the deletion... ocsp check can for sure be a bottleneck and single point of failure... and the pattern of my problems could be compatible with that.. but the cache should help, not cause the problem.... I don't know how you implement SSL/TLS, but I remember I have read Apple doesn't like OpenSSL, if it can help....

**Comment by ncw** [comment]
@baro77 rclone uses the go TLS implementation which was written from scratch in go.

I don't really understand what is going on - whether there is a problem with the TLS library with OS X.  

It might be that this issue is relevant: https://github.com/golang/go/issues/14514

In which case compiling rclone yourself would fix the problem.  I'd do it for you but I don't have a mac :-(

If you want to have a try compiling rclone, then you'll need to install `go` and `git`
  * [install go](https://golang.org/doc/install)
  * [install git](https://git-scm.com/download/mac)
  * run go `get -u github.com/ncw/rclone`
  * find the rclone binary in ~/go/bin/rclone

**Comment by baro77** [comment]
@ncw I guess your explanation has led me to the right route.... Known that golang environment takes care of the whole TLS handshake, I checked https://drive.amazonaws.com/drive/v1/account/endpoint with a browser to inspect the certificate.... TA-DA! Damn Avast local antivirus's Web Shield breaking HTTPS with its own certificate  >:-|    So sorry to haven't thought about it earlier stated that I'm an IT :( ... now I have disabled Avast Web Shield, next weekend I'll use rclone a lot, so see you in 3/4 days to confirm the problem is eventually over...

**Comment by ncw** [comment]
@baro77  I should probably re-open the go bug (or create a new issue) as it is definitely the same problem.

Let me know how it goes!

**Comment by baro77** [comment]
@ncw bad news, it re-happened, so you were right about it not being related to any dirty way Avast antivirus uses/registers its own certs.... This weekend i have done about 1000 "rclone delete" and I can say the problem occuring-rate falls between 3% and 6%, without any specific pattern.
It isn't a real concern while using rclone inside script.. in bash for example you can always do a 
`until <rclone command>; do sleep 1; done`
to repeat the command if it fails (in my tests the problem rarely appeared also in the second iteration of the loop, and never 3 times); of course it can cause much more problems with rclone mount... (but I guess you already implemented retryings massively there).
what I really cannot understand is: stated that it's a cert validation problem, the casuality and rate of occurrence seems inconsistent with how a load balancing front-end usually works (because I mean... ok, sometimes go fails to validate the cert, but the actual cause must be someone changing the returned cert...) ... i need to try to record exchanged certs and compare them with success/fail log....

**Comment by ncw** [comment]
It does seem very strange that it only happens some of the time.

Maybe someone is intermittently man-in-the-middling your connection?

Or maybe there is a mis-configured server in the cluster with a dud certificate.

If you do with `-vv --dump-headers` likely you will see which server it is actually connecting to.

**Comment by baro77** [comment]
Hi @ncw ! Here they are to calls to "rclone lsd" (-v because I'm still using 1.35):

`alpha1:~ baro$ rclone lsd -v --dump-headers ACD:
2017/02/27 12:47:32 rclone: Version "v1.35" starting with parameters ["rclone" "lsd" "-v" "--dump-headers" "ACD:"]
2017/02/27 12:47:32 >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
2017/02/27 12:47:32 HTTP REQUEST (req 0xc4201ec780)
2017/02/27 12:47:32 GET /drive/v1/account/endpoint HTTP/1.1
Host: drive.amazonaws.com
User-Agent: rclone/v1.35
Authorization: XXXX
Accept-Encoding: gzip

2017/02/27 12:47:32 >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
2017/02/27 12:47:34 <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
2017/02/27 12:47:34 HTTP RESPONSE (req 0xc4201ec780)
2017/02/27 12:47:34 HTTP/1.1 200 OK
Cache-Control: no-store
Connection: keep-alive
Content-Type: application/json
Date: Mon, 27 Feb 2017 11:47:34 GMT
Pragma: no-cache
Server: Server
Vary: User-Agent
X-Amzn-Requestid: 870ca58c-fce2-11e6-93d7-ab3f8d892ffc

2017/02/27 12:47:34 <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
2017/02/27 12:47:34 >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
2017/02/27 12:47:34 HTTP REQUEST (req 0xc4201ec2d0)
2017/02/27 12:47:34 GET /drive/v1/nodes?filters=kind%3AFOLDER+AND+isRoot%3Atrue HTTP/1.1
Host: cdws.eu-west-1.amazonaws.com
User-Agent: rclone/v1.35
Authorization: XXXX
Accept-Encoding: gzip

2017/02/27 12:47:34 >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
2017/02/27 12:47:34 <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
2017/02/27 12:47:34 HTTP RESPONSE (req 0xc4201ec2d0)
2017/02/27 12:47:34 HTTP/1.1 200 OK
Cache-Control: no-store
Connection: keep-alive
Content-Type: application/json
Date: Mon, 27 Feb 2017 11:47:34 GMT
Pragma: no-cache
Server: Server
Vary: User-Agent
X-Amzn-Requestid: 872fbd4c-fce2-11e6-9929-0b0b002b2edb

2017/02/27 12:47:34 <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
2017/02/27 12:47:34 amazon drive root '': Modify window not supported
2017/02/27 12:47:34 amazon drive root '': Reading ""
2017/02/27 12:47:34 >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
2017/02/27 12:47:34 HTTP REQUEST (req 0xc4201ec960)
2017/02/27 12:47:34 GET /drive/v1/nodes?filters=parents%3ATYqgPs9tTACsly7qg9mNWA HTTP/1.1
Host: cdws.eu-west-1.amazonaws.com
User-Agent: rclone/v1.35
Authorization: XXXX
Accept-Encoding: gzip

2017/02/27 12:47:34 >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
2017/02/27 12:47:34 <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
2017/02/27 12:47:34 HTTP RESPONSE (req 0xc4201ec960)
2017/02/27 12:47:34 HTTP/1.1 200 OK
Cache-Control: no-store
Connection: keep-alive
Content-Type: application/json
Date: Mon, 27 Feb 2017 11:47:34 GMT
Pragma: no-cache
Server: Server
Vary: User-Agent
X-Amzn-Requestid: 873d7989-fce2-11e6-b5e3-df6a1468ded8

2017/02/27 12:47:34 <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
2017/02/27 12:47:34 amazon drive root '': Finished reading ""
          -1 2012-09-27 07:19:03        -1 Video
          -1 2012-09-27 07:19:03        -1 Documenti
          -1 2012-09-27 07:19:03        -1 Immagini
          -1 2017-02-01 22:14:01        -1 Pictures
          -1 2017-02-15 16:45:38        -1 rclone
2017/02/27 12:47:34 Go routines at exit 8





alpha1:~ baro$ rclone lsd -v --dump-headers ACD:
2017/02/27 12:48:55 rclone: Version "v1.35" starting with parameters ["rclone" "lsd" "-v" "--dump-headers" "ACD:"]
2017/02/27 12:48:55 >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
2017/02/27 12:48:55 HTTP REQUEST (req 0xc4200e4a50)
2017/02/27 12:48:55 GET /drive/v1/account/endpoint HTTP/1.1
Host: drive.amazonaws.com
User-Agent: rclone/v1.35
Authorization: XXXX
Accept-Encoding: gzip

2017/02/27 12:48:55 >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
2017/02/27 12:48:57 <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
2017/02/27 12:48:57 HTTP RESPONSE (req 0xc4200e4a50)
2017/02/27 12:48:57 Error: x509: certificate signed by unknown authority
2017/02/27 12:48:57 <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
2017/02/27 12:48:57 Failed to create file system for "ACD:": failed to get endpoints: Get https://drive.amazonaws.com/drive/v1/account/endpoint: x509: certificate signed by unknown authority

`

As you can see, the first one was ok, the second not. And you can see that when all is good drive.amazonaws.com is resolved in cdws.eu-west-1.amazonaws.com (confirmed by a shell "host drive.amazonaws.com" command as well) 
BUT when things get wrong, the problem appears quite soon, just when contacting the fronted "drive.amazonaws.com"... so I used the following script to check returned certs chains for multiple accesses to "drive.amazonaws.com":

`until false; do (echo "EOF\n\n" | openssl s_client -showcerts -connect drive.amazonaws.com:443 | sed "/BEGIN.*CERTIFICATE/,/END.*CERTIFICATE/d"  >> provaSSL.txt) 2>> provaSSL.txt; echo "----ALTRA----" >> provaSSL.txt; sleep 1; done`

the until loop create an infinite loop; the openssl s_client open a ssl/tls session; the sed cancel the noise created in the output by base64-encoded certs: the redirects capture both stout and stderr in the file "provaSSL.txt. 
2412 iteration, always the same certs chain :( (I expected differences to justify errors I get sometimes with rclone)... so it seems no bad certs appearing occasionally...

**Comment by ncw** [comment]
Great debugging.

This seems most likely that it is an intermittent problem with rclone contacting Avast.  Can you turn it off and confirm that?

Also can you try building rclone from source (see above) and try that too?

**Comment by baro77** [comment]
Hi @ncw 
First of all let me say Avast actually isn't playing any role: since  i discovered its man-on-the-middle web shield, I have immediately deactivated it... anyway to be estra-sure I uninstalled it yesterday, as you asked for: no changes in what is happening, so no Avast's fault.

I have built rclone from source and it really seems resolving the problem, but i haven't had a lot of time yet to stress my test procedures with the new rclone binary, so I'll confirm it next days.

But I can already add something... I was a bit lazy about installing git and go, just because i wasn't too confident, or maybe because as a professional I'm too protocols-centric/biased ;) ... so before compiling rclone, I obtained CRLs and OCSPs URLs from certs chain and I poisoned their DNS resolution via my /etc/hosts, to get them unreachable... if the problem was in that stage of certs evaluation, my problem-occurence pattern would increase from 3..6% to 100%. But that hasn't happened... so the problem is with validation of self-signed root cert (the only one without CRL or OCSP).... and in my opinion that's really weird by GO's ssl/tls subsystem:  first because its behaviour isn't always the same, second because the root cert is a public one (and of course known to macOS's keychain: "System Roots / Certificates" - keychains are the place where certs and credentials are stored in macOS)... so [golang/go#14514](https://github.com/golang/go/issues/14514) isn't strictly relevant, at least regarding the part of user's certs/local PKIs VS public global ones...

I hope my points are useful. Thanks, see ya

**Comment by ncw** [comment]
How did it go - did an rclone built from source fix the problem?

**Comment by baro77** [comment]
Hi @ncw 
sorry for late update, but I have been very busy this week.. and I thought you too with 1.36 release, so I didn't hurry up writing to you... 
I have done around 400 rclone single operations on my remote with rclone built on my mac, and the problem has never appeared (I was expecting 10...20 errors).
it's still a mystery for me why the error pattern with "standard" rclone is intermittent...
I can add to the previous analysis that's not a problem with recent macOSes' System Integrity Protection (I tried to disable it)

**Comment by ncw** [comment]
That is good to know that rclone built on the mac is OK.

That probably means that something like golang/go#14514 is relevant.  If the native build is OK but the cross compiled build isn't.

I don't know why it should be intermittent though - that is strange!

**Comment by baro77** [comment]
Ok ... I will deal again with this issue when and if I'll have any useful news...
Btw... can you give me any tip about how to build a specific rclone version with git+go `get -u github.com/ncw/rclone` command? (I mean, I could be interested in self-building current stable release and not the last commited one -  now for example v1.35 instead of 1.35.99....) Thanks!

**Comment by ncw** [comment]
To build a specific version

  cd $GOPATH/src/github.com/ncw/rclone
  git checkout v1.35
  go install

should do it

**Comment by baro77** [comment]
Sorry but those commands work neither with v1.35 nor with v1.36: after checking-out, on "go install" I get:
can't load package: package github.com/ncw/rclone: no buildable Go source files in /Users/baro/go/src/github.com/ncw/rclone

**Comment by ncw** [comment]
After you've done the `go get` try this commands above

```
cd /Users/baro/go/src/github.com/ncw/rclone
git checkout v1.36
go build
```

That should build a binary in the current directory

**Comment by diamondsw** [comment]
Afraid I haven't much to offer other than "OS X me too", and no form of anti-virus or other software that would interfere with an SSL handshake. If I ever have a free moment I'll see if I can compile rclone and run tests as well (but don't hold your breath - two toddlers in my house).

**Comment by felixbuenemann** [comment]
I am seeing this error sometimes with google drive when I start a copy, it always works if I retry:

```
Failed to create file system for "drive:foo/bar": couldn't read info about Drive: Get https://www.googleapis.com/drive/v2/about?alt=json: x509: certificate signed by unknown authority
```

Would it be possible to classify this as a low level error, so it gets retried automatically or at least once? 

It's strange that this error always appears at the start of an operation and never in the middle.

I have a direct connection via ipv6 or nat for ipv4, no antivirus, proxy or similar, rclone v1.3.5 on maOS 10.12.4.

**Comment by baro77** [comment]
Hi @ncw 
Sorry for late reply, but they are busy days.
Ok, now I get it:
after `go get -u`, `go build`  creates the executable in $GOPATH/src/github.com/ncw/rclone , while `go install` creates it in $GOPATH/bin, but only if it is empty 
I want to underline only that checking-out v1.36 the resulting rclone answers like this:
`./rclone -V
rclone v1.36`
instead checking-out v1.35 i still get the "DEV":
`./rclone -V
rclone v1.35-DEV`
(but maybe #1041 applies only to newest releases)

**Comment by ncw** [comment]
@brandur - yes #1041 will only apply to 1.36 and later!

**Comment by felixbuenemann** [comment]
This is very weird. I now have one terminal session on macOS 10.12 where this is consistently happening, also with the latest beta build 250.

If I build rclone myself from master, the connection simply hangs, which is the behavior I reported in  #1362.

However, if I open a new terminal window, rclone works fine…

I diffed the output of printenv in both sessions, but didn't see anything suspicious, the differences are mostly stuff like DISPLAY and SSH_AUTH_SOCK.

**Comment by felixbuenemann** [comment]
OK, I made a simple sample program that connects to the same host:

```go
package main

import (
	"crypto/tls"
	"log"
)

func main() {
	log.SetFlags(log.Lshortfile)

	conf := &tls.Config{
	//InsecureSkipVerify: true,
	}

	conn, err := tls.Dial("tcp", "www.googleapis.com:443", conf)
	if err != nil {
		log.Println(err)
		return
	}
	defer conn.Close()

	n, err := conn.Write([]byte("GET /drive/v2/about?alt=json HTTP/1.0\r\nHost: www.googleapis.com\r\n\r\n"))
	if err != nil {
		log.Println(n, err)
		return
	}

	buf := make([]byte, 1024)
	n, err = conn.Read(buf)
	if err != nil {
		log.Println(n, err)
		return
	}

	println(string(buf[:n]))
}
```

If I run it in a fresh terminal window with `go run foo.go` it works fine, but in the "broken" session, the name resolution fails:

```
go run foo.go                                                                                                                                              127
foo.go:17: dial tcp: lookup www.googleapis.com: no such host
```

**Comment by felixbuenemann** [comment]
Some more info about the broken session that might be useful: I had a long running rclone process running in it just fine and then I disconnected my external monitor and when I logged back in it took longer than usual and it restored windows, so I think the window server must have crashed. I then accidentally disconnected the wrong usb drive (the one rclone was accessing, which caused rclone to abort). After reconnecting the drive and restarting rclone the above mentioned errors started in that session.
